packages/mobile/src/identity/feelessVerification.ts

```diff
@@ -0,0 +1,1061 @@
+        pepperQuotaRemaining,
+      },
+    })
+  )
+}
+
+function* fetchVerifiedMtw(contractKit: ContractKit, walletAddress: string, e164Number: string) {
+  Logger.debug(TAG, '@fetchVerifiedMtw', 'Starting fetch')
+  const feelessVerificationState: FeelessVerificationState = yield select(
+    feelessVerificationStateSelector
+  )
+
+  const phonHashDetails: PhoneNumberHashDetails = yield call(
+    fetchPhoneHashDetailsFromCache,
+    e164Number
+  )
+  const { phoneHash } = phonHashDetails
+  const lookupResult: IdentifierLookupResult = yield call(lookupAttestationIdentifiers, [phoneHash])
+  const possibleMtwAddresses = getAddressesFromLookupResult(lookupResult, phoneHash)
+
+  if (!possibleMtwAddresses) {
+    Logger.debug(TAG, '@fetchVerifiedMtw', 'No verified MTW found')
+    return null
+  }
+
+  // TODO: Incorporate fix for the error on this and remove try/catch block
+  let verificationResults: Array<Result<true, WalletValidationError>>
+  try {
+    verificationResults = yield all(
+      possibleMtwAddresses.map((possibleMtwAddress) =>
+        call(
+          verifyWallet,
+          contractKit,
+          possibleMtwAddress,
+          ALLOWED_MTW_IMPLEMENTATIONS,
+          walletAddress
+        )
+      )
+    )
+  } catch (error) {
+    Logger.debug(TAG, '@fetchVerifiedMtw', 'Unable to validate existing verified accounts')
+    Logger.error(TAG, '@fetchVerifiedMtw', error)
+    return null
+  }
+
+  const verifiedMtwAddresses = possibleMtwAddresses.filter(
+    (address, i) => verificationResults[i].ok
+  )
+
+  if (verifiedMtwAddresses.length > 2) {
+    throw Error(
+      'More than one verified MTW with walletAddress as signer found. Should never happen'
+    )
+  }
+
+  const verifiedMtwAddress = verifiedMtwAddresses[0]
+
+  if (!verifiedMtwAddress) {
+    Logger.debug(TAG, '@fetchVerifiedMtw', 'No verified MTW found')
+    return null
+  }
+
+  yield put(
+    feelessUpdateVerificationState({
+      ...feelessVerificationState,
+      status: {
+        isVerified: true,
+        numAttestationsRemaining: 0,
+        total: NUM_ATTESTATIONS_REQUIRED,
+        completed: NUM_ATTESTATIONS_REQUIRED,
+      },
+      komenci: {
+        ...feelessVerificationState.komenci,
+        unverifiedMtwAddress: verifiedMtwAddress,
+      },
+    })
+  )
+
+  return verifiedMtwAddress
+}
+
+function* reFetchVerifiedMtw(
+  contractKit: ContractKit,
+  walletAddress: string,
+  e164Number: string,
+  startingPepperQuota: number
+) {
+  Logger.debug(TAG, '@reFetchVerifiedMtw', 'Starting fetch')
+  const feelessVerificationState: FeelessVerificationState = yield select(
+    feelessVerificationStateSelector
+  )
+
+  const { pepperQuotaRemaining } = feelessVerificationState.komenci
+  // If Komenci was used to fetch their pepper, check if the they have a verified
+  // account. If user already had their pepper, then check would have happened
+  // in `feelessFetchVerificationState`
+  if (startingPepperQuota > pepperQuotaRemaining) {
```

Comment:
Instead of assuming a function was called previously, can we record somthing in the state to let us know for sure? This check just seems fragile.

Reply:
Yeah you're right. That's a good suggestion I'll do that.